File: apps/posting/posting_core/clients/linkedin.py
from __future__ import annotations

import time
import urllib.parse

from posting_core.http_client import HttpRequestError, request, request_json
from posting_core.publish_config import LINKEDIN_ACCESS_TOKEN, LINKEDIN_AUTHOR_URN, log
from posting_core.text import strip_leading_emojis

def upload_linkedin_binary(upload_url, file_path):
    with open(file_path, "rb") as f:
        data = f.read()
    resp = request(
        upload_url,
        data=data,
        headers={"Content-Type": "application/octet-stream"},
        method="PUT",
        timeout=60,
    )
    return resp.status == 200 or resp.status == 201


def upload_linkedin_image(local_path):
    init_payload = {
        "initializeUploadRequest": {
            "owner": LINKEDIN_AUTHOR_URN
        }
    }
    res = call_linkedin("rest/images?action=initializeUpload", init_payload)
    value = res.get("value") or {}
    upload_url = value.get("uploadUrl")
    if not upload_url:
        upload_instructions = value.get("uploadInstructions") or []
        if upload_instructions:
            upload_url = upload_instructions[0].get("uploadUrl")
    media_urn = value["image"]
    if not upload_url:
        raise Exception(f"LinkedIn image initializeUpload did not return uploadUrl: {value}")

    log("Uploading image binary to LinkedIn...")
    success = upload_linkedin_binary(upload_url, local_path)
    if not success:
        raise Exception("Failed to upload binary to LinkedIn")
    return media_urn


def upload_linkedin_chunk(upload_url, chunk_data):
    try:
        resp = request(
            upload_url,
            data=chunk_data,
            headers={"Content-Type": "application/octet-stream"},
            method="PUT",
            timeout=60,
        )
    except HttpRequestError as err:
        log(f"HTTPError uploading chunk: {err.status} {err.reason}. Response body: {err.body}")
        raise Exception(f"LinkedIn Chunk Upload HTTP {err.status}: {err.body}")
    etag = resp.headers.get("ETag") or resp.headers.get("etag")
    if not etag:
        raise Exception("ETag header not found in chunk upload response")
    return etag


def call_linkedin(endpoint, payload, method="POST"):
    url = f"https://api.linkedin.com/{endpoint}"
    headers = {
        "Authorization": f"Bearer {LINKEDIN_ACCESS_TOKEN}",
        "Linkedin-Version": "202606",
        "X-Restli-Protocol-Version": "2.0.0",
        "Content-Type": "application/json"
    }
    
    try:
        return request_json(
            url,
            method=method,
            payload=payload,
            headers=headers,
            timeout=30,
            empty_id_header="x-restli-id",
        )
    except HttpRequestError as err:
        log(f"HTTPError in call_linkedin: {err.status} {err.reason}. Response body: {err.body}")
        raise Exception(f"LinkedIn API HTTP {err.status}: {err.body}")


def publish_to_linkedin(text, media_items):
    if not LINKEDIN_ACCESS_TOKEN or not LINKEDIN_AUTHOR_URN:
        return None
        
    log("Publishing to LinkedIn...")
    text = strip_leading_emojis(text)
    try:
        payload = {
            "author": LINKEDIN_AUTHOR_URN,
            "commentary": text,
            "visibility": "PUBLIC",
            "distribution": {
                "feedDistribution": "MAIN_FEED",
            },
            "lifecycleState": "PUBLISHED"
        }
        
        if len(media_items) > 0:
            image_items = [item for item in media_items if item["type"] == "IMAGE"]
            video_items = [item for item in media_items if item["type"] == "VIDEO"]
            media_urn = None

            if len(image_items) >= 2 and not video_items:
                image_items = image_items[:20]
                log(f"Uploading {len(image_items)} images to LinkedIn multiImage post...")
                image_urns = [upload_linkedin_image(item["local_path"]) for item in image_items]
                payload["content"] = {
                    "multiImage": {
                        "images": [
                            {"id": image_urn, "altText": "Post image"}
                            for image_urn in image_urns
                        ]
                    }
                }
            elif len(media_items) > 1:
                log("LinkedIn mixed/video multi-media is not supported yet; publishing the first media item only.")

            item = media_items[0]
            item_type = item["type"]
            local_path = item["local_path"]
            
            if payload.get("content", {}).get("multiImage"):
                pass
            elif item_type == "VIDEO":
                init_payload = {
                    "initializeUploadRequest": {
                        "owner": LINKEDIN_AUTHOR_URN,
                        "fileSizeBytes": local_path.stat().st_size,
                        "uploadCaptions": False,
                        "uploadThumbnail": False
                    }
                }
                res = call_linkedin("rest/videos?action=initializeUpload", init_payload)
                
                upload_token = res["value"]["uploadToken"]
                media_urn = res["value"]["video"]
                upload_instructions = res["value"]["uploadInstructions"]
                
                log(f"Uploading video in {len(upload_instructions)} chunks to LinkedIn...")
                uploaded_parts = []
                
                with open(local_path, "rb") as f:
                    for i, instruction in enumerate(upload_instructions):
                        upload_url = instruction["uploadUrl"]
                        first_byte = instruction["firstByte"]
                        last_byte = instruction["lastByte"]
                        
                        f.seek(first_byte)
                        chunk_size = last_byte - first_byte + 1
                        chunk_data = f.read(chunk_size)
                        
                        log(f"Uploading chunk {i+1}/{len(upload_instructions)} ({len(chunk_data)} bytes)...")
                        etag = upload_linkedin_chunk(upload_url, chunk_data)
                        uploaded_parts.append(etag)
                
                log("Finalizing video upload on LinkedIn...")
                finalize_payload = {
                    "finalizeUploadRequest": {
                        "video": media_urn,
                        "uploadToken": upload_token,
                        "uploadedPartIds": uploaded_parts
                    }
                }
                call_linkedin("rest/videos?action=finalizeUpload", finalize_payload)
                
                log("Waiting for video to become AVAILABLE on LinkedIn...")
                encoded_urn = urllib.parse.quote(media_urn)
                deadline = time.monotonic() + 300
                while True:
                    if time.monotonic() >= deadline:
                        raise Exception("Timed out waiting for LinkedIn video validation")
                    
                    status_res = call_linkedin(f"rest/videos/{encoded_urn}", payload=None, method="GET")
                    status = status_res.get("status")
                    log(f"LinkedIn video status: {status}")
                    if status == "AVAILABLE":
                        break
                    elif status == "PROCESSING_FAILED":
                        raise Exception("LinkedIn video processing failed")
                    time.sleep(10)
            else:
                media_urn = upload_linkedin_image(local_path)
                
            if "content" not in payload:
                payload["content"] = {
                    "media": {
                        "title": "Post Media",
                        "id": media_urn
                    }
                }
            
        res = call_linkedin("rest/posts", payload)
        log(f"LinkedIn post published successfully URN: {res.get('id', 'Unknown')}")
        return res.get("id")
        
    except Exception as exc:
        log(f"Error publishing to LinkedIn: {exc}")
        return None

File: apps/backend/src/social/linkedin.ts
import fs from "node:fs";
import type { BackendConfig } from "../config.js";
import type { PublishResult } from "../queue/errors.js";
import { HttpPublishError } from "../queue/errors.js";
import { type PublishMediaItem, payloadMedia, payloadText, stripLeadingEmojis } from "./payload.js";

type LinkedInResponse = Record<string, unknown> & { id?: string; value?: Record<string, unknown>; status?: string };

export async function publishToLinkedIn(
  payload: Record<string, unknown>,
  config: BackendConfig,
  fetchImpl: typeof fetch = fetch,
): Promise<PublishResult> {
  if (!config.LINKEDIN_ACCESS_TOKEN || !config.LINKEDIN_AUTHOR_URN) throw new Error("missing LinkedIn credentials");
  const media = payloadMedia(payload).filter((item) => item.localPath);
  const post: Record<string, unknown> = {
    author: config.LINKEDIN_AUTHOR_URN,
    commentary: stripLeadingEmojis(payloadText(payload)),
    visibility: "PUBLIC",
    distribution: { feedDistribution: "MAIN_FEED" },
    lifecycleState: "PUBLISHED",
  };

  const images = media.filter((item) => item.type === "IMAGE");
  const videos = media.filter((item) => item.type === "VIDEO");
  if (images.length >= 2 && videos.length === 0) {
    const urns = await Promise.all(images.slice(0, 20).map((item) => uploadImage(item.localPath!, config, fetchImpl)));
    post.content = { multiImage: { images: urns.map((id) => ({ id, altText: "Post image" })) } };
  } else if (media[0]) {
    const mediaUrn =
      media[0].type === "VIDEO"
        ? await uploadVideo(media[0], config, fetchImpl)
        : await uploadImage(media[0].localPath!, config, fetchImpl);
    post.content = { media: { title: "Post Media", id: mediaUrn } };
  }

  const result = await callLinkedIn(config, "rest/posts", fetchImpl, { method: "POST", body: JSON.stringify(post) });
  return { ok: Boolean(result.id), id: result.id ?? null, raw: result };
}

async function uploadImage(localPath: string, config: BackendConfig, fetchImpl: typeof fetch): Promise<string> {
  const initialized = await callLinkedIn(config, "rest/images?action=initializeUpload", fetchImpl, {
    method: "POST",
    body: JSON.stringify({ initializeUploadRequest: { owner: config.LINKEDIN_AUTHOR_URN } }),
  });
  const value = initialized.value ?? {};
  const instructions = Array.isArray(value.uploadInstructions) ? (value.uploadInstructions as Array<Record<string, unknown>>) : [];
  const uploadUrl = stringValue(value.uploadUrl) || stringValue(instructions[0]?.uploadUrl);
  const imageUrn = stringValue(value.image);
  if (!uploadUrl || !imageUrn) throw new Error(`LinkedIn image initializeUpload incomplete: ${JSON.stringify(value)}`);
  await uploadBinary(uploadUrl, await fs.promises.readFile(localPath), fetchImpl);
  return imageUrn;
}

async function uploadVideo(item: PublishMediaItem, config: BackendConfig, fetchImpl: typeof fetch): Promise<string> {
  const localPath = item.localPath!;
  const initialized = await callLinkedIn(config, "rest/videos?action=initializeUpload", fetchImpl, {
    method: "POST",
    body: JSON.stringify({
      initializeUploadRequest: {
        owner: config.LINKEDIN_AUTHOR_URN,
        fileSizeBytes: fs.statSync(localPath).size,
        uploadCaptions: false,
        uploadThumbnail: false,
      },
    }),
  });
  const value = initialized.value ?? {};
  const videoUrn = stringValue(value.video);
  const uploadToken = stringValue(value.uploadToken);
  const instructions = Array.isArray(value.uploadInstructions) ? (value.uploadInstructions as Array<Record<string, unknown>>) : [];
  if (!videoUrn || !uploadToken || instructions.length === 0)
    throw new Error(`LinkedIn video initializeUpload incomplete: ${JSON.stringify(value)}`);

  const file = await fs.promises.open(localPath, "r");
  const uploadedPartIds: string[] = [];
  try {
    for (const instruction of instructions) {
      const firstByte = Number(instruction.firstByte);
      const lastByte = Number(instruction.lastByte);
      const buffer = Buffer.alloc(lastByte - firstByte + 1);
      await file.read(buffer, 0, buffer.length, firstByte);
      uploadedPartIds.push(await uploadBinary(stringValue(instruction.uploadUrl), buffer, fetchImpl, true));
    }
  } finally {
    await file.close();
  }
  await callLinkedIn(config, "rest/videos?action=finalizeUpload", fetchImpl, {
    method: "POST",
    body: JSON.stringify({ finalizeUploadRequest: { video: videoUrn, uploadToken, uploadedPartIds } }),
  });

  const deadline = Date.now() + 300_000;
  while (Date.now() < deadline) {
    const status = await callLinkedIn(config, `rest/videos/${encodeURIComponent(videoUrn)}`, fetchImpl, { method: "GET" });
    if (status.status === "AVAILABLE") return videoUrn;
    if (status.status === "PROCESSING_FAILED") throw new Error("LinkedIn video processing failed");
    await delay(10_000);
  }
  throw new Error("Timed out waiting for LinkedIn video validation");
}

async function uploadBinary(url: string, bytes: Buffer, fetchImpl: typeof fetch, requireEtag = false): Promise<string> {
  if (!url) throw new Error("LinkedIn upload URL is missing");
  const bodyBytes = Uint8Array.from(bytes).buffer;
  const response = await fetchImpl(url, { method: "PUT", headers: { "Content-Type": "application/octet-stream" }, body: bodyBytes });
  const body = await response.text();
  if (!response.ok) throw new HttpPublishError(`LinkedIn upload ${response.status}: ${body}`, response.status, body);
  const etag = response.headers.get("etag") ?? "";
  if (requireEtag && !etag) throw new Error("ETag header not found in LinkedIn chunk upload response");
  return etag;
}

async function callLinkedIn(
  config: BackendConfig,
  endpoint: string,
  fetchImpl: typeof fetch,
  init: RequestInit,
): Promise<LinkedInResponse> {
  const response = await fetchImpl(`https://api.linkedin.com/${endpoint}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${config.LINKEDIN_ACCESS_TOKEN}`,
      "Linkedin-Version": config.LINKEDIN_API_VERSION,
      "X-Restli-Protocol-Version": "2.0.0",
      "Content-Type": "application/json",
      ...init.headers,
    },
  });
  const body = await response.text();
  if (!response.ok) throw new HttpPublishError(`LinkedIn API ${response.status}: ${body}`, response.status, body);
  const parsed = body ? (JSON.parse(body) as LinkedInResponse) : {};
  return { ...parsed, ...(parsed.id ? {} : response.headers.get("x-restli-id") ? { id: response.headers.get("x-restli-id")! } : {}) };
}

function stringValue(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
